builtins: Implement ST_GeneratePoints function

mknycha opened the pull request:
This function generates pseudo-random points until the requested number are found within the input area.

Release note (sql change): Implement geo builtin ST_GeneratePoints

Resolves: #48942

Dependent on: https://github.com/cockroachdb/cockroach/pull/54476

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/58288)
<!-- Reviewable:end -->

mknycha commented on pkg/geo/geomfn/generate_points.go:
```diff
@@ -0,0 +1,227 @@
+	bboxWidth := bbox.HiX - bbox.LoX
+	bboxHeight := bbox.HiY - bbox.LoY
+	bboxArea := bboxHeight * bboxWidth
+	if area == 0.0 || bboxArea == 0.0 {
```
That's how it works in PostGIS, so I wanted to double-check - Can the `bboxArea` be zero when `area` is not zero?

mknycha replied:
All right, I removed `bboxArea` check here.

mknycha commented on pkg/geo/geomfn/generate_points.go:
```diff
@@ -0,0 +1,227 @@
+				return nil, errors.Newf("could not set X and Y for coord sequence: %v", err)
+			}
+			defer func() {
+				// err = geos.CoordSequenceDestroy(gseq)
```
1) I know that `gseq` should be destroyed, but when I uncomment this line I get a really weird error and can't figure out why:
```
fatal error: unexpected signal during runtime execution
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x5fd1210]

runtime stack:
runtime.throw(0x4a2e35f, 0x2a)
	/usr/local/Cellar/go/1.15.5/libexec/src/runtime/panic.go:1116 +0x72
runtime.sigpanic()
	/usr/local/Cellar/go/1.15.5/libexec/src/runtime/signal_unix.go:726 +0x48c
```
I compared the C API handler code to the one for `PreparedGeometry`, I have noticed that the destroy method for `CoordSequence` has no `const` argument type:
```
extern void GEOS_DLL GEOSPreparedGeom_destroy_r(GEOSContextHandle_t handle,
                                                const GEOSPreparedGeometry* g);

extern void GEOS_DLL GEOSCoordSeq_destroy_r(GEOSContextHandle_t handle,
                                            GEOSCoordSequence* s);
```
Could you help to debug it, please? 🙏 
2) Can you recommend any better way to handle errors in the `defer` statement?

mknycha replied:
Still on (1) - sorry, but how do I do that? I have tried to add a new function like this:
```
CR_GEOS_Status CR_GEOS_GeometryDestroy(CR_GEOS* lib, CR_GEOS_Slice a) {
  std::string error;
  auto handle = initHandleWithErrorBuffer(lib, &error);

  auto wkbReader = lib->GEOSWKBReader_create_r(handle);
  auto geomA = lib->GEOSWKBReader_read_r(handle, wkbReader, a.data, a.len);
  lib->GEOSWKBReader_destroy_r(handle, wkbReader);
  if (geomA != nullptr) {
    lib->GEOSGeom_destroy_r(handle, geomA);
  }
  lib->GEOS_finish_r(handle);
  return toGEOSString(error.data(), error.length());
}
```
But in the end, I still got the same error.
I could make this function expect a `CR_GEOS_Geometry` type input but in that case, I would probably need another way of keeping a pointer to `CR_GEOS_Geometry` GeometryPoint and transforming it to a `CR_GEOS_Slice` (so that it can be used in `PrepareGeometry`).

otan replied:
I think you don't need to call `GEOSCoordSeq_destroy_r` at all, since `GeometryCreate` makes it the owner of the Seq object.

I think we should avoid doing this GeometryCreate altogether. Can we just create a Point using `go-geom` and use that EWKB?

Also for simplicity, I'm happy for you to just use `geos.Intersects` instead of `PreparedGeometry`.

mknycha replied:
Ah, that makes sense. I will go with a simple point then 👍

mknycha commented on pkg/geo/geomfn/generate_points.go:
```diff
@@ -0,0 +1,227 @@
+			defer func() {
+				// err = geos.CoordSequenceDestroy(gseq)
+			}()
+			gpt, err := geos.GeometryPointCreate(gseq)
```
Does `gpt` need to be destroyed explicitly?

otan replied:
I would think so.

mknycha reviewed (commented):
Opening as a draft for now, I have several items I would like to discuss.

otan commented on pkg/geo/geos/geos.cc:
```diff
@@ -1160,6 +1189,126 @@ CR_GEOS_Status CR_GEOS_HausdorffDistanceDensify(CR_GEOS* lib, CR_GEOS_Slice a, C
+  if (geom != nullptr) {
+    *ret = lib->GEOSPrepare_r(handle, geom);
+    // TODO: make sure underlying geom is freed too.
+    // lib->GEOSGeom_destroy_r(handle, geom);
```
this needs to return a struct, so that we free the `GEOSGeometry` underneath.

Consider something like:

```
struct PreparedGeometry {
  CR_GEOS_Geometry *g;
   CR_GEOS_PreparedInternalGeometry *p; // this corresponds to PreparedGeometry in GEOS.
}

CR_GEOS_Status CR_GEOS_PrepareGeometry(CR_GEOS* lib, CR_GEOS_Slice a, CR_GEOS_PreparedGeometry* ret) {
    ....
    *ret = {.g = geom, .p = lib->GEOSPrepare_r(handle, geom)}
}
```

So that in `CR_GEOS_PreparedGeometryDestroy`, we call `GEOSGeom_destroy_r` on `g` as well.

mknycha replied:
Ok, I changed the `PreparedGeometry` structure accordingly.
Please just note that in the updated Go code `geos.PrepareGeometry` on error returns an empty `PreparedGeometry` instead of `nil`, I hope I got it right.

otan reviewed (commented):
Sorry for not getting back to you sooner - new year break.
Looking good, responded to your help comments!

mknycha commented on pkg/geo/geomfn/generate_points_test.go:
```diff
@@ -0,0 +1,121 @@
+		},
+		{
+			"Polygon with a specified SRID",
+			args{geo.MustParseGeometry("SRID=4326;POLYGON((0 0,2 5,2.5 4,3 5,3 1,0 0))"), 5, 1996},
```
Good catch, added

mknycha commented on pkg/geo/geomfn/generate_points.go:
```diff
@@ -0,0 +1,228 @@
+	return out, nil
+}
+
+func polygonGeometryToPoints(g geo.Geometry, npoints int, seed int64) (*geom.MultiPoint, error) {
```
Fixed

mknycha commented on pkg/geo/geomfn/generate_points.go:
```diff
@@ -0,0 +1,228 @@
+	return results, nil
+}
+
+func multipolygonGeometryToPoints(
```
Fixed

mknycha commented on pkg/geo/geomfn/generate_points.go:
```diff
@@ -0,0 +1,228 @@
+	}
+	empty, err := IsEmpty(g)
+	if err != nil {
+		return geo.Geometry{}, errors.Newf("could not check if Geometry is empty: %v", err)
```
Fixed

mknycha commented on pkg/geo/geomfn/generate_points.go:
```diff
@@ -0,0 +1,228 @@
+	mpt.SetSRID(int(srid))
+	out, err := geo.MakeGeometryFromGeomT(mpt)
+	if err != nil {
+		return geo.Geometry{}, errors.Newf("could not transform geom.T into Geometry: %v", err)
```
Fixed

mknycha commented on pkg/geo/geomfn/generate_points.go:
```diff
@@ -0,0 +1,228 @@
+	case geopb.ShapeType_MultiPolygon:
+		generatePointsFunction = multipolygonGeometryToPoints
+	default:
+		return geo.Geometry{}, errors.New("unsupported type")
```
Fixed

mknycha commented on pkg/geo/geomfn/generate_points.go:
```diff
@@ -0,0 +1,228 @@
+// that are covered by the geometry provided.
+// npoints is the number of points to return.
+// seed is the seed used by the random numbers generator.
+func geometryToPoints(g geo.Geometry, npoints int, seed int64) (geo.Geometry, error) {
```
Fixed

mknycha commented on pkg/geo/geomfn/generate_points.go:
```diff
@@ -0,0 +1,228 @@
+)
+
+// GeneratePoints generates provided number of pseudo-random points for the input area.
+func GeneratePoints(g geo.Geometry, npoints int, seed int64) (geo.Geometry, error) {
```
Makes sense, fixed

mknycha commented on pkg/geo/geomfn/generate_points.go:
```diff
@@ -0,0 +1,228 @@
+		return nil, errors.New("zero area input Polygon")
+	}
+	// Gross up our test set a bit to increase odds of getting coverage in one pass.
+	sampleNpoints := float64(npoints) * bboxArea / area
```
Fixed

mknycha commented on pkg/geo/geomfn/generate_points.go:
```diff
@@ -0,0 +1,228 @@
+	for i := 0; i < gmp.NumPolygons(); i++ {
+		poly := gmp.Polygon(i)
+		subarea := poly.Area()
+		subNpoints := int(math.Round(float64(npoints) * subarea / area))
```
Ok, changed

mknycha commented on pkg/geo/geomfn/generate_points.go:
```diff
@@ -0,0 +1,228 @@
+)
+
+// GeneratePoints generates provided number of pseudo-random points for the input area.
+func GeneratePoints(g geo.Geometry, npoints int, seed int64) (geo.Geometry, error) {
```
Changed

mknycha commented on pkg/geo/geomfn/generate_points.go:
```diff
@@ -0,0 +1,228 @@
+		return nil, errors.Newf("could not prepare Geometry: %v", err)
+	}
+	defer func() {
+		err = geos.PreparedGeomDestroy(gprep)
```
Fair enough, fixed

otan commented on pkg/geo/geos/geos.cc:
```diff
@@ -1160,6 +1173,66 @@ CR_GEOS_Status CR_GEOS_HausdorffDistanceDensify(CR_GEOS* lib, CR_GEOS_Slice a, C
+// PreparedGeometry
+//
+
+CR_GEOS_Status CR_GEOS_PrepareGeometry(CR_GEOS* lib, CR_GEOS_Slice a, CR_GEOS_PreparedGeometry* ret) {
```
can we call this `CR_GEOS_Prepare` to mimic the GEOS name

mknycha replied:
Sure, changed

mknycha commented on pkg/geo/geomfn/generate_points.go:
```diff
@@ -0,0 +1,217 @@
+				}
+				gpt, err := geo.MakeGeometryFromPointCoords(x, y)
+				if err != nil {
+					return nil, errors.Newf("could not create geometry Point: %v", err)
```
Sure, fixed

mknycha commented on pkg/geo/geomfn/generate_points.go:
```diff
@@ -0,0 +1,217 @@
+					}
+				}
+			}
+			if nPointsGenerated == nPoints || iterations > 100 {
```
I was not sure, so used the PostGIS value. I think it's just an arbitrary number to avoid going through too many point generations if they are not covered by the geometry.
Yes, fixed.

otan commented on pkg/geo/geomfn/generate_points.go:
```diff
@@ -0,0 +1,214 @@
+		// Generate points and test them.
+		nPointsGenerated := 0
+		iterations := 0
+		for nPointsGenerated < nPoints || iterations <= 100 {
```
i'm still not sure why 100 is here. This `<=` would also give 101 iterations; we could do `for nPointsGenerated, iterations := 0, 0; nPointsGenerated < nPoints || iterations <= 100; iterations++` as well
Do we stop generating points after 100? Do we really need to cap this? What if nPoints > 100? Should the cap be 10 * nPoints instead?

mknycha replied:
I played around with different geometries, I think I understand your concern now.
For each `iteration` here we are still going through each cell in the grid, for which we are generating a point. The purpose of limiting `iterations` is to avoid going through the grid over and over if there is a little likelihood of generating a point that will intersect with our geometry.
However, from the above, you can see that the real complexity here is `iterations * len(cells)`. The number of cells is calculated from `sampleNPoints := float64(nPoints) * bboxArea / area` so, the more points we are supposed to generate and the higher `bounding box area / geometry area` ratio, the more complex calculations.
For `sampleNPoints` of ~14000000 it took about 18 seconds to generate points on my machine.

I suggest we get rid of `iterations` cap at all. If we need a complexity limitation, keeping in mind differences in execution time between different machines, I can make the code throw a timeout error after some time. What's your opinion?

otan replied:
not a big fan of timeouts as they are non deterministic. if we need something like this (not opposed but this has to be clear in the function `Info` to the user) i'd rather it be a function of the nPoints as input.

mknycha replied:
Fair enough, so you'd prefer to keep the condition, but instead of `100` use `nPoints * 10`?
I do not have much experience with optimization like this, I just thought that in theory, a user running CRDB on a very fast computer could want to handle much more complexity. This solution sounds good, too.

otan replied:
Yeah prefer the latter if we decide we need this.

otan commented on pkg/geo/geomfn/generate_points.go:
```diff
@@ -0,0 +1,214 @@
+	}
+	var sampleHeight, sampleWidth int
+	var sampleCellSize float64
+	// Calculate the grids we're going to randomize within.
```
This needs some commenting for the next person to look at this. What is a grid in this instance, and what are sample heights / widths? Maybe a hollistic view of the algorithm

mknycha replied:
Ok, added function description.

otan reviewed (commented):
Thanks! I think I still need some comments to understand the algorithm, still confused about the iterations bit.

mknycha commented on pkg/geo/geos/geos.go:
```diff
@@ -59,6 +59,9 @@ var geosOnce struct {
 }
 
+// PreparedGeometry is an instance of a GEOS PreparedGeometry.
+type PreparedGeometry C.CR_GEOS_PreparedGeometry
```
I have tried to edit it that way and I have changed the related functions:
```
CR_GEOS_Status CR_GEOS_Prepare(CR_GEOS* lib, CR_GEOS_Slice a, CR_GEOS_PreparedGeometry** ret) {
  std::string error;
  auto handle = initHandleWithErrorBuffer(lib, &error);

  auto wkbReader = lib->GEOSWKBReader_create_r(handle);
  auto geom = lib->GEOSWKBReader_read_r(handle, wkbReader, a.data, a.len);
  lib->GEOSWKBReader_destroy_r(handle, wkbReader);
  if (geom != nullptr) {
    auto preparedGeom = lib->GEOSPrepare_r(handle, geom);
    CR_GEOS_PreparedGeometry tmp = {.g = geom, .p = preparedGeom};
    CR_GEOS_PreparedGeometry *tmp2 = &tmp;
    *ret = tmp2;
  }
  lib->GEOS_finish_r(handle);
  return toGEOSString(error.data(), error.length());
}

CR_GEOS_Status CR_GEOS_PreparedGeometryDestroy(CR_GEOS* lib, CR_GEOS_PreparedGeometry* a) {
  std::string error;
  auto handle = initHandleWithErrorBuffer(lib, &error);
  lib->GEOSPreparedGeom_destroy_r(handle, a->p);
  lib->GEOSGeom_destroy_r(handle, a->g);
  lib->GEOS_finish_r(handle);
  return toGEOSString(error.data(), error.length());
}

CR_GEOS_Status CR_GEOS_PreparedIntersects(CR_GEOS* lib, CR_GEOS_PreparedGeometry* a, CR_GEOS_Slice b, char* ret) {
  return CR_GEOS_PreparedBinaryPredicate(lib, lib->GEOSPreparedIntersects_r, a->p, b, ret);
}
```
I managed to get no compilation errors, but an execution one:
```
fatal error: unexpected signal during runtime execution
[signal SIGSEGV: segmentation violation code=0x1 addr=0x1 pc=0x5fd5861]

runtime stack:
runtime.throw(0x4a2ecae, 0x2a)
	/usr/local/Cellar/go/1.15.5/libexec/src/runtime/panic.go:1116 +0x72
runtime.sigpanic()
	/usr/local/Cellar/go/1.15.5/libexec/src/runtime/signal_unix.go:726 +0x48c

goroutine 15 [syscall]:
runtime.cgocall(0x48ae000, 0xc000080928, 0x4010100)
	/usr/local/Cellar/go/1.15.5/libexec/src/runtime/cgocall.go:133 +0x5b fp=0xc0000808e0 sp=0xc0000808a8 pc=0x400875b
github.com/cockroachdb/cockroach/pkg/geo/geos._Cfunc_CR_GEOS_PreparedIntersects(0xd004190, 0x7ffeefbff348, 0xc000013600, 0x15, 0xc0000af1c0, 0x0, 0x0)
	_cgo_gotypes.go:724 +0x4d fp=0xc000080928 sp=0xc0000808e0 pc=0x47acfed
github.com/cockroachdb/cockroach/pkg/geo/geos.PreparedIntersects.func1(0xd004190, 0x7ffeefbff348, 0xc000013600, 0x15, 0x40, 0xc0000af1c0, 0x0, 0x0)
	/Users/marcinknychala/go/src/github.com/cockroachdb/cockroach/pkg/geo/geos/geos.go:766 +0xc5 fp=0xc000080980 sp=0xc000080928 pc=0x47b8c65
```
Could you help me out, please?
Should I somehow explicitly allocate memory for this CR_GEOS_PreparedGeometry?
I can also change `ST_GeneratePoints` to work without `PreparedGeometry`, but I really want to make sense out of it and deliver something optimized :)

otan replied:
You would need to `malloc` space for `tmp` beforehand I think:

```
*ret = new CR_GEOS_PreparedGeometry(); // or define initializer on this method
ret->g = ...
ret -> p = ....
```

and to free the pointer in `CR_GEOS_PreparedGeometryDestroy`.

> I can also change ST_GeneratePoints to work without PreparedGeometry, but I really want to make sense out of it and deliver something optimized :)

Of course!

mknycha replied:
Thanks for the explanation, it's working now!

mknycha commented on pkg/geo/geomfn/generate_points_test.go:
```diff
@@ -0,0 +1,123 @@
+		},
+		{
+			"empty geometry",
+			args{geo.MustParseGeometry("POLYGON EMPTY"), 4, 1},
```
I have already added a test like that 🤔

otan replied:
missed it - apologies

mknycha commented on pkg/geo/geomfn/generate_points_test.go:
```diff
@@ -0,0 +1,123 @@
+		name    string
+		args    args
+		want    geo.Geometry
+		wantErr bool
```
Good idea, changed

otan reviewed (approved):
looks awesome! thanks for all the hard work you've put into this, and going through the hard work to get it nice :) 

bors r+

otan reviewed (commented):
bors r+

otan commented:
bors r-

looks like you need to run `make bazel-generate`. mind doing that?

mknycha commented:
> looks like you need to run `make bazel-generate`. mind doing that?

Not at all

mknycha committed:
builtins: Implement ST_GeneratePoints function
Release note (sql change): Implement geo builtin ST_GeneratePoints

mknycha commented:
I am sorry @miretskiy, I uploaded too many changes by an accident causing additional reviewers to be assigned. Fixed now, please disregard.

@otan Please note that in the last push I also removed assigning NULL to a deleted pointer `a = NULL;` that used to be [here](https://github.com/cockroachdb/cockroach/pull/58288/files#diff-08b20894e5f643a614f62b3db326a49fc5192a1ae92953e2eea759bcf6039093R1204). I think it will be ok as long as `a` is not used anymore, besides I believe it should be handled one level abstraction higher.

otan reviewed (approved):
lgtm

bors r+
